Cubical/Experiments/truncations.agda

```diff
@@ -0,0 +1,165 @@
+    L = L1 (proof x) (proof y)
+
+data setTrunc (A : Set) : Set where
+  tinc : A → setTrunc A
+  tIsSet : isSet (setTrunc A)
+
+setTruncElim : (A : Set) (B : (setTrunc A) → Set)
+               (bS : (x : setTrunc A) → isSet (B x))
+               (f : (x : A) → B (tinc x))
+               (x : setTrunc A) → B x
+setTruncElim A B bS f = proof
+  where
+  proof : (x : setTrunc A) → B x
+  proof (tinc x) = f x
+  proof (tIsSet x y p q i j) = L i j
+    where
+    sA = setTrunc A
+    C = tIsSet x y p q
+    M1 : {x y : sA} (p : x ≡ y) → B x → B y → Set
+    M1 p a b = PathP (λ i → B (p i)) a b
+    M2 : {x y : sA} {p q : x ≡ y} (r : p ≡ q) {a : B x} {b : B y}
+       → (M1 p a b) → (M1 q a b) → Set
+    M2 r {a = a} {b = b} c d = PathP (λ i → M1 (r i) a b) c d
+    B' : {x y : sA} {p q : x ≡ y} (r : p ≡ q) → Set
+    B' {x} {y} {p} {q} r = {a : B x} {b : B y} (c : M1 p a b) (d : M1 q a b)
+                           → M2 r c d
+    L0 : B' (λ i j → x)
+    L0 c d = bS x _ _ c d 
+    B0 B1 : I → Set
+    B0 k = B' (λ i j → C i0 (j ∧ k))
+    B1 k = B' (λ i j → C (i ∧ k) j)
+    L1 = transp B0 i0 L0
+    L2 = transp B1 i0 L1
+    L = L2 (λ i → proof (p i)) (λ i → proof (q i))
+
+data groupoidTrunc (A : Set) : Set where
+  tinc : A → groupoidTrunc A
+  tIsGroupoid : isGroupoid (groupoidTrunc A)
+
+groupoidTruncElim : (A : Set) (B : (groupoidTrunc A) → Set)
+                    (bG : (x : groupoidTrunc A) → isGroupoid (B x))
+                    (f : (x : A) → B (tinc x)) (x : groupoidTrunc A) → B x
+groupoidTruncElim A B bG f = proof
+  where
+  proof : (x : groupoidTrunc A) → B x
+  proof (tinc x) = f x
+  proof (tIsGroupoid x y p q r s i j k) = L i j k
+    where
+    gA = groupoidTrunc A
+    C = tIsGroupoid x y p q r s
+    M1 : {x y : gA} (p : x ≡ y) → B x → B y → Set
+    M1 p a b = PathP (λ i → B (p i)) a b
+    M2 : {x y : gA} {p q : x ≡ y} (r : p ≡ q) {a : B x} {b : B y}
+       → (M1 p a b) → (M1 q a b) → Set
+    M2 r {a = a} {b = b} c d = PathP (λ i → M1 (r i) a b) c d
+    M3 : {x y : gA} {p q : x ≡ y} {r s : p ≡ q} (u : r ≡ s)
+         {a : B x} {b : B y} {c : M1 p a b} {d : M1 q a b}
+         → (M2 r c d) → (M2 s c d) → Set
+    M3 u {c = c} {d = d} e f = PathP (λ i → M2 (u i) c d) e f
+    B' : {x y : gA} {p q : x ≡ y} {r s : p ≡ q} → (r ≡ s) → Set
+    B' {x} {y} {p} {q} {r} {s} u =
+      {a : B x} {b : B y} {c : M1 p a b} {d : M1 q a b}
+      (e : M2 r c d) (f : M2 s c d) → M3 u e f
+    L0 : B' (λ i j k → x)
+    L0 e f = bG x _ _ _ _ e f
+    B0 B1 B2 : I → Set
+    B0 i = B' (λ j0 j1 j2 → C i0 i0 (i ∧ j2))
+    B1 i = B' (λ j0 j1 j2 → C i0 (i ∧ j1) j2)
+    B2 i = B' (λ j0 j1 j2 → C (i ∧ j0) j1 j2)
+    L1 = transp B0 i0 L0
+    L2 = transp B1 i0 L1
+    L3 = transp B2 i0 L2
+    L = L3 (λ i j → proof (r i j)) (λ i j → proof (s i j))
+
+data g2Trunc (A : Set) : Set where
+  g2inc : A → g2Trunc A
+  g2IsTwoGroupoid : isTwoGroupoid (g2Trunc A)
+
+g2TruncElim : (A : Set) (B : (g2Trunc A) → Set)
```

Comment:
Can you compress this proof as well? I want to understand how it compares to:

https://github.com/mortberg/cubicaltt/blob/pi4s3_dimclosures/examples/brunerie2.ctt#L2087